packages/config-helpers/src/define-config.js

```diff
@@ -0,0 +1,380 @@
+	}
+
+	result.name = `${baseConfigName} > ${extensionName}`;
+
+	return result;
+}
+
+/**
+ * Processes a list of extends elements.
+ * @param {ConfigWithExtends} config The config object.
+ * @param {WeakMap<Config, string>} configNames The map of config objects to their names.
+ * @return {Config[]} The flattened list of config objects.
+ */
+function processExtends(config, configNames) {
+	if (!config.extends) {
+		return [config];
+	}
+
+	const {
+		/** @type {Config[]} */
+		extends: extendsList,
+
+		/** @type {Config} */
+		...configObject
+	} = config;
+
+	const extensionNames = new WeakMap();
+
+	// replace strings with the actual configs
+	const objectExtends = extendsList.map(extendsElement => {
+		if (typeof extendsElement === "string") {
+			const pluginConfig = findPluginConfig(config, extendsElement);
+
+			// assign names
+			if (Array.isArray(pluginConfig)) {
+				pluginConfig.forEach((pluginConfigElement, index) => {
+					extensionNames.set(
+						pluginConfigElement,
+						`${extendsElement}[${index}]`,
+					);
+				});
+			} else {
+				extensionNames.set(pluginConfig, extendsElement);
+			}
+
+			return pluginConfig;
+		}
+
+		return /** @type {Config} */ (extendsElement);
+	});
+
+	const result = [];
+
+	for (const { indexPath, value: extendsElement } of flatTraverse(
+		objectExtends,
+	)) {
+		const extension = /** @type {Config} */ (extendsElement);
+		const baseConfigName = /** @type {string} */ (configNames.get(config));
+		const extensionName =
+			extensionNames.get(extendsElement) ??
+			getExtensionName(extendsElement, indexPath);
+
+		result.push(
+			extendConfig(
+				configObject,
+				baseConfigName,
+				extension,
+				extensionName,
+			),
+		);
+	}
+
+	// original config last
+	result.push(configObject);
+
+	return result.flat();
+}
+
+/**
+ * Processes a list of config objects and arrays.
+ * @param {ConfigWithExtends[]} configList The list of config objects and arrays.
+ * @param {WeakMap<Config, string>} configNames The map of config objects to their names.
+ * @return {Config[]} The flattened list of config objects.
+ */
+function processConfigList(configList, configNames) {
+	return configList.flatMap(config => processExtends(config, configNames));
+}
+
+//-----------------------------------------------------------------------------
+// Exports
+//-----------------------------------------------------------------------------
+
+/**
+ * Helper function to define a config array.
+ * @param {ConfigWithExtendsArray} args The arguments to the function.
+ * @returns {Config[]} The config array.
+ */
+export function defineConfig(...args) {
+	const configNames = new WeakMap();
+	const configs = [];
+
+	// first flatten the list of configs and get the names
+	for (const { indexPath, value } of flatTraverse(args)) {
+		const config = /** @type {ConfigWithExtends} */ (value);
+
+		// save config name for easy reference later
+		configNames.set(config, getConfigName(config, indexPath));
```

Comment:
Here I think it would be good to check if `value` is an object. Currently, for example, `defineConfig(null)` fails with `TypeError: Cannot read properties of null (reading 'name')`.